charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,200 @@
+
+    resource_class_kind = resource_class.__name__
+    if replicas == 1:
+        logger.info(f"{resource_class_kind}/{resource_name} readyReplicas == {replicas}")
+    else:
+        logger.info(
+            f"{resource_class_kind}/{resource_name} readyReplicas == {replicas} (waiting for '1')"
+        )
+
+    assert replicas == 1, f"Waited too long for {resource_class_kind}/{resource_name}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # verify that all created CRDs and Services in namespace are deleted
+    for rsc in [
+        CustomResourceDefinition,
+        Service,
+    ]:
+        rsc_list = lightkube_client.list(
+            rsc,
+            labels=[("app.juju.is/created-by", APP_NAME)],
+            namespace=ops_test.model_name,
+        )
+        assert not list(rsc_list)
+
+    # verify that all created Pods in namespace are deleted
+    for rsc in [
+        Pod,
+    ]:
+        rsc_list = lightkube_client.list(
+            rsc,
+            labels=[("app", TENSORBOARD_NAME)],
+            namespace=ops_test.model_name,
+        )
+        assert not list(rsc_list)
+
+    # verify that the PVC and Tensorboard are deleted
+    try:
+        for rsc, name in [
+            (PersistentVolumeClaim, PVC_NAME),
+            (TENSORBOARD_RESOURCE, TENSORBOARD_NAME),
+        ]:
+            _ = lightkube_client.get(rsc, name=name, namespace=ops_test.model_name)
+    except HTTPStatusError:
+        assert True
```

Comment:
Can you explain this exception catch?

Reply:
I found this [here](https://github.com/canonical/notebook-operators/blob/188d8073495a2d934a1c4fdf8937f8edc6949c90/charms/jupyter-controller/tests/integration/test_charm.py#L270-L271) and initially thought it was redundant. I started testing expecting just the `ApiError` and most of the times it was fine. But I did bump into the occasion where this error was raised instead.

Reply:
hmm interesting.  @i-chvets do you recall what this is about?  

Feels like we should at least say to the reader why `assert True` here is an ok thing

Reply:
So, I realised what's happening here. When lightkube hits the K8s API it expects an `httpx.HTTPStatusError` in case of failure, which it then tries to transform using `transform_exception`:
```python
def transform_exception(e: httpx.HTTPError):
    if isinstance(e, httpx.HTTPStatusError) and e.response.headers['Content-Type'] == 'application/json':
        return ApiError(request=e.request, response=e.response)
    return e
```
* When the CRD doesn't exist, the lightkube client gets a 404 `HTTPStatusError` trying to hit the endpoint which it raises directly, since the content type is `text/plain; charset=utf-8`.
* On the other hand, when the CRD exists, the API server response is indeed of `application/json` content type, so the exception is wrapped in an `ApiError`. 

At the very least we should log helpful messages.

Reply:
There's a bigger issue here, however.

For starters, since `ApiError` inherits from `HTTPStatusError`, the first `except` acts as an umbrella. That would not be so terrible under different circumstances (it'd just render `except ApiError` redundant), however in our case it means that we're not really testing what we think we are here:
* As explained above, when we try to fetch a resource, the corresponding CRD of which has been deleted, a 404 `HTTPStatusError` exception is raised
* However, we already know that the CRD has been deleted (we previously check explicitly)
* This means that we just catch the `HTTPStatusError` raised here and never check whether the `Tensorboard` object itself has been deleted
  * The above is something that is not in our power anyway, since we rely on K8s to remove any dependents after we delete the CRD

Reply:
I think it is safe to just check if we've deleted the CRD and not worry about checking CRs.  If k8s deletes a CRD without removing the CRs, there's a much bigger problem going on.